projects/super_res/model/autoreg_diffusion_mod.py

```diff
@@ -0,0 +1,1442 @@
+import os
+import math
+from pathlib import Path
+from random import random
+from functools import partial
+from collections import namedtuple
+from joblib import Parallel, delayed
+
+import numpy as np
+
+import torch
+from torch import nn, einsum
+import torch.nn.functional as F
+import wandb
+
+import piq
+
+from kornia import filters
+from torch.optim import Adam
+
+from einops import rearrange, reduce
+from einops.layers.torch import Rearrange
+
+from PIL import Image
+
+import matplotlib as mpl
+from matplotlib.cm import ScalarMappable as smap
+
+from tqdm.auto import tqdm
+from ema_pytorch import EMA
+
+import flow_vis
+
+from accelerate import Accelerator
+
+from .network_swinir import SwinIR as net
+
+# constants
+
+ModelPrediction =  namedtuple('ModelPrediction', ['pred_noise', 'pred_x_start'])
+
+# helpers functions
+
+def calculate_crps(truth, pred, num_samples, num_videos_per_batch, num_frames, img_channels, img_size):
+    truth_cdf = np.zeros((256, 1, num_videos_per_batch, num_frames, img_channels, img_size, img_size), dtype = 'uint8')
+    for i in range(256):
+        truth_cdf[i, :, :, :, :, :, :] = (truth <= i).astype('uint8')
+    pred_cdf = np.zeros((256, num_samples, 1, num_videos_per_batch, num_frames, img_channels, img_size, img_size), dtype = 'uint8')
+    for j in range(256):
+        pred_cdf[j, :, :, :, :, :, :, :] = (pred <= j).astype('uint8')
+    red_pred_cdf = pred_cdf.mean(1)
+    temp = np.square(red_pred_cdf - truth_cdf)
+    temp_dz = temp.sum(0)
+    temp_dz_dd = temp_dz.mean(axis = (3, 4, 5))
+    temp_dz_dd_dt = temp_dz_dd.mean(2)
+    return temp_dz_dd_dt.mean()
+
+def save_image(tensor, path):
+    im = Image.fromarray((tensor[:,:,:3] * 255).astype(np.uint8))
+    im.save(path)
+    return None
+
+def exists(x):
+    return x is not None
+
+def default(val, d):
+    if exists(val):
+        return val
+    return d() if callable(d) else d
+
+def identity(t, *args, **kwargs):
+    return t
+
+def cycle(dl):
+    while True:
+        for data in dl:
+            yield data
+
+def has_int_squareroot(num):
+    return (math.sqrt(num) ** 2) == num
+
+def num_to_groups(num, divisor):
+    groups = num // divisor
+    remainder = num % divisor
+    arr = [divisor] * groups
+    if remainder > 0:
+        arr.append(remainder)
+    return arr
+
+def convert_image_to_fn(img_type, image):
+    if image.mode != img_type:
+        return image.convert(img_type)
+    return image
+
+# normalization functions
+
+def normalize_to_neg_one_to_one(img):
+    return img * 2 - 1
+
+def unnormalize_to_zero_to_one(t):
+    return (t + 1) * 0.5
+
+# ssf modules
+
+def gaussian_pyramids(input, base_sigma = 1, m = 5):
+    
```

Comment:
nit: I think it would make sense to remove this spacing. Maybe it's copied from another codebase where this is standard? Still, it doesn't seem to be standard in many codebases and IMHO makes it hard to read.

Reply:
Aah of course